subsys: fs/nvs: fix write block size issues

aurel32 opened the pull request:
This patch series is an alternative to #7600 to fix write block size issues. It ensures that all writes to the flash are done in multiple of the write block size, even if the data to write is smaller, by adding some padding. It also make sure that the flash read and write match the destination or source buffer to avoid stack corruption or data leakage.

This fixes #7671 for me, running on a nucleo_l432kc board (8-byte write block, smaller writes not possible).

@Laczen, note that it only fixes the write block size issues, I guess there are more things to extract from your patch, at least the tests and associated documentation.

@MaureenHelm, could you please try it out on your frdm_k64 platform? Thanks.

aurel32 committed:
subsys: fs/nvs: kill a warning

GCC complains that last_entry.len and last_entry.data_addr might be
uninitialized in _nvs_gc. Fix that.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 committed:
subsys: fs/nvs: never read more than the destination buffer

Reading more than the destination buffer means that data is overwritten
possibly on the stack. This causes unpredictable behaviours like
crashes.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 committed:
subsys: fs/nvs: never write more than the source buffer

Writing more than the source buffer means that some random data,
possibly coming from the stack, ends-up in the flash. This could be
a security issue.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 committed:
subsys: fs/nvs: explicitly pad flash writes up to the write block size

Some SoCs do not allow shorter writes than the write block size, usually
when they have ECC memory. This patch first write all data in multiple
of the write block size and then do a last write with the data padded.
It uses 0xff as the padding byte to avoid wearing-out the flash.

nvs_append_close() is slightly modified to compute the crc16 that will
be put in the slot over the size defined in the header, to match the way
it is checked on read.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 committed:
subsys: fs/nvs: remove explicit padding from structures

Now that the flash writes are padded up to the write block size, there
is no need to have explicit padding fields in the _nvs_sector_hdr and
_nvs_data_slt structure. This allow to save space when the write block
size equals to 1 or 2

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

Laczen commented:
@aurel32, Thanks for the PR, this is better than my approach.

Laczen reviewed (approved):
Nice fix, thanks.

aurel32 commented:
I think the bus fault seen by @MaureenHelm on the frdm_k64f platform is due to the wrong assumption that the flash is mapped at address 0. This trick is used in the NVS code to compute the CRC16 when reading or writing data. Note that it is even more wrong with an SPI flash.

I have pushed an additional patch to fix that.

Laczen commented:
@aurel32, agreed the calculation of the crc16 directly on flash would better be avoided especially for spi flash. It is not really needed to include the pads in the crc16 calculation. The closing could also be done by reading from flash, calculating CRC, writing seal - then there is no need to pass the data.

aurel32 commented:
@Laczen, I have done the computation on the RAM copy of the data, as I believed it might be faster than reloading the data from the flash. That said given the usual flash memory write time, the difference is probably not be noticeable. That would allow to use the same CRC computation code for both read and write.

I can implement that, but it would be nice to have the opinion of someone else on that aspect first.

Laczen commented:
@aurel32, I would do it on the flash data. The main function of the slt is to close the write. In theory this could be done much later than the write of the data. The idea of the CRC is also not so much to say that the data is correct but more that the slot is written correct.

MaureenHelm commented:
> This was the first test I have written and it seems I need to understand the use of the yamls better.

Have a look at `doc/subsystems/test/sanitycheck.rst`.

aurel32 committed:
subsys: fs/nvs: fix _nvs_sector_is_used

_nvs_sector_is_used() never uses the offset argument. As a consequence,
it only check the first sector of the flash. Fix that.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 committed:
subsys: fs/nvs: do not assume the flash is mapped at address 0

The two functions that compute the crc16 when writing (nvs_append_close)
and when reading (nvs_check_crc) currently assume that the flash is
also mapped in read mode at address 0. This is not true on all SoCs, and
even less on an SPI flash.

Fix this by adding a new nvs_compute_crc() function which compute the
CRC16 of an entry using the flash using nvs_flash_read, in blocks of
write_block_size. This might not be the optimal size, but it keeps the
stack usage small.

Use this function in both nvs_append_close() and nvs_check_crc() instead
of accessing the flash from address 0.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 commented:
> a. I think the pads need to be reinserted in the structures: this would allow future changes to use the pads without breaking the data already in flash.

Do you mean dropping the patch `subsys: fs/nvs: remove explicit padding from structures`? I agree that having a way to extend the format might be a good idea. The alternative is to add a `version` field to the _nvs_sector_hdr, so that the format can be extended later.

> b. The data_slt CRC needs to be calculated on the flash. This of course limits the meaning of the CRC: it only means the slot was closed properly.

Ok. I have changed the commit to compute the CRC from the flash in both cases, adding a new `nvs_compute_crc()` function for that.

> A later extension could be to add an extra CRC to the data_slt to include the CRC on the data, than data correctness can easily be checked by comparing the two CRC's.

I agree that it could be a later extension. Now, I am not sure it's the best to write the two CRC to the flash, as soon as the error is detected (i.e. before closing the slot), the error can be returned to the application.

Two additional notes:
- I noticed that nvs_write also assume that the flash is mapped at address 0 when CONFIG_NVS_PROTECT_FLASH=y. I'll try to work on a fix later, no time today.
- I added a patch to fix `_nvs_sector_is_used`. Without it the flash is not always properly initialized.

aurel32 commented:
> Hi @aurel32, your remarks and code corrections (thanks again for this) have made me think about the basics of NVS again. I am thinking doing the following changes:

I am new to the zephyr development, but isn't that a bit late in the release cycle to change the filesystem format? That said it won't be too later after the release.

> a. Remove the data_slt

Do you mean data_hdr will contain all the metadata? How do you plan to write things in that case? first writing the data leaving space for the header, which is always written after the data?

> b. Extend the data_hdr to include a CRC and a reverse or previous len (plen) to allow stepping back in NVS. At the moment there is no check that the data_hdr is correctly written and this is not good. Adding a plen would improve the reading performance and garbage collection.

It's always good to have a CRC, so if it's not anymore in data_slt, it should indeed be in data_hdr. Adding a previous len entry looks indeed an excellent idea to optimize the algorithms. It might even simplify a lot of code as most of the time we want to read the most recent version of a value (or the nth more recent).

> c. Extend the data with a CRC, calculated and written together with the data.

I don't really get what you mean by "extend the data" with a CRC?

> What is your opinion about this ? I am doubting about adding a version field to the sector_hdr as this would mean for each read the current sector header would need to be read to decide on how to handle the data. Garbage collection would then also need to do the conversion, ...

ok, fair enough

Laczen commented:
@aurel32, I was proposing to many changes at once. Let me rephrase the first step:

Instead of doing the CRC check in the data_slt, the CRC check is moved to every flash write/read. All data that is written to flash has a CRC, including sector_hdr, data_hdr and data. As everything in flash has a CRC the data_slt becomes obsolete.

The next step, adding a previous len should not be done at the same time, I will keep that for later.

lemrey commented:
Hi @aurel32, @Laczen I have a few comments.

As I understand, an entry's slot (its presence) gives us the guarantee that no power loss has occurred during the write operation. This guarantee is given just by the slot being found in the flash: CRC are not necessary for this.

The CRC can be used by NVS to determine other kind of errors:
- data in flash being overwritten

And additionally:
- application buffer being modified while being written to flash
- hardware failure

The last two I put separately because, from what I understand, NVS computes the CRC from a RAM copy of the data in flash (we read the flash contents into RAM). Wouldn't it be better to compute the CRC on the application buffer instead (am I missing something obvious)? Not only it would be faster, but it would also enable NVS to detect the case when said buffer was modified while being written or when the hardware failed to write altogether (optionally). The CRC could be kept in one of the parameters to nvs_append() or calculated in its entirety when using nvs_write().

I would suggest that if only one CRC is used (as we have now), then it should be calculated over the header as well (id-len), since the header might be corrupted as well (hardware failure or overwritten).

Also, for an application using NVS, it should be trivial to bundle a CRC, checksum or hash together with the data and ckeck for errors when reading, if necessary. In that case, NVS could be relieved of detecting data errors on the data itself, which could be costly and not desirable or redundant in some circumstances.

aurel32 commented:
> Instead of doing the CRC check in the data_slt, the CRC check is moved to every flash write/read. All data that is written to flash has a CRC, including sector_hdr, data_hdr and data. As everything in flash has a CRC the data_slt becomes obsolete.

What do you mean for every write? You consider the data part as a whole with a single CRC, or regular CRC among the data every X bytes of flash.

Just adding a single CRC at the end just look like a the current slot to me.

aurel32 commented:
> The last two I put separately because, from what I understand, NVS computes the CRC from a RAM copy of the dataThe last two I put separately because, from what I understand, NVS computes the CRC from a RAM copy of the data in flash (we read the flash contents into RAM). Wouldn't it be better to compute the CRC on the application buffer instead (am I missing something obvious)? Not only it would be faster, but it would also enable NVS to detect the case when said buffer was modified while being written or when the hardware failed to write altogether (optionally). The CRC could be kept in one of the parameters to nvs_append() or calculated in its entirety when using nvs_write(). in flash (we read the flash contents into RAM). Wouldn't it be better to compute the CRC on the application buffer instead (am I missing something obvious)? Not only it would be faster, but it would also enable NVS to detect the case when said buffer was modified while being written or when the hardware failed to write altogether (optionally). The CRC could be kept in one of the parameters to nvs_append() or calculated in its entirety when using nvs_write().

That's what I have done in one of the earlier version of the patchset, but we agreed that so far we should leave the behavior unchanged. This causes some complexity as the CRC should include the padding, which is not available in the RAM copy.

aurel32 commented:
> A combined CRC on entry slot (data_hdr) and data is therefore not a good idea. I am at the moment working on a version (on top of the changes by aurel32) where all items written to flash have there own CRC (only a 8 bit CRC to reduce redundancy and load), and all CRC's are calculated on RAM data.

I think care should be taken in not multiplying the number of small entries in the flash. The tendency is to get bigger and bigger write block size. 8 bytes is not an uncommon value, so writing an 1-byte entry can take up to 8-bytes in the flash.

In that regard, I wonder if a single header containing the following entries to be enough:
- ID
- len
- CRC of the header
- CRC of the data (therefore computed from RAM)

We need the CRC of the header separated from the CRC of the data to have to way to verify if the header is valid. It might not be valid in case of a power failure. The CRC of the data also allows to check for the validity of the data, and detect a power failure. Ideally the CRC should be computed from RAM and from flash just after the data have been written to make sure the header and data have been correctly written. The failure can be returned to the application layer which can handle the failure.

lemrey commented:
Hi @Laczen , sorry perhaps I wasn't clear enough. I was referring to the _nvs_data_slt data type.
My assertion is that if an entry has a _nvs_data_slt written in flash, because _nvs_data_slt is written last (it seals the entry) then the entry has been written in its _entirety_ i.e. there was no power loss. This can be determined without a CRC check.

With the addition of a CRC we become able to determine if anything has been overwritten in flash, and depending on which buffer we compute the CRC from (RAM or flash) we are also able to

- detect  hardware errors and
- detect that the entry data (in the RAM buffer) has changed while being written to flash

These two classes of errors are write-time errors, as opposed to the classic CRC read-time error.
I understand that computing the CRC over the RAM copy can be problematic now, that's perfectly fair.
I hadn't thought about the padding issues :) My point wasn't much for the sake of speed though, but for the additional error dection capabilities that it makes possible.

In any case, I am convinced now that keeping two CRC, one for _nvs_data_hdr and one for the entry data gives us the benefit of being able to detect when a _nvs_data_hdr is corrupted _separately_ from the entry's data being corrupted. This is a good thing, since if a _nvs_data_hdr then we have a critical failure, as opposed to just corrupt data of one entry.

Are there any thoughts as to how to recover from a corrupt _nvs_data_hdr scenario?

nvlsianpu reviewed (approved):
@aurel32, @Laczen Thanks for this patch and excellent fixing messages story.
@lemrey remarks are valuable as well, we can introduce the second crc in the future.

Laczen commented:
@aurel32, @nvlsianpu, I have made some changes on top of this PR. I have pushed in to my github repo (https://github.com/Laczen/zephyr) as branch nvs_pr7674. I have tried to create small code increments and commented all of these in the hope that they would show up the same way as @aurel32 makes changes, of course this did not work. I have no idea how to integrate my changes to the PR.

The main changes are:

-  CRC's are added when writing to flash, checked when reading, they are no longer part of any structure.
- All items stored have CRC's.
- The data_slt structure is removed.

I have tested this for fs->write_block_size=1, 4 and 8.

@aurel32, could you please test on your system. @nvlsianpu, whats the best way to create a PR on the zephyr tree , so that it can be reviewed ?

aurel32 commented:
Hi @Laczen, thanks for your improvements on it. To answer quickly your questions about git and PR, in general it's better to do small commits that you can reorganize and merge before sending a PR. Also it's possible to send patches written by someone else in a PR.

To come back about your patchset, as it changes things that I have introduced in my patchset I have just looked at it as a big diff between master and your nvs_pr7674. That's probably how things should be submitted at this point, a single commit.

The way you bundled the CRC within the padding if possible is very nicely done. The only drawback I see is that you can't read the flash if you don't know the write block size, but I think it's not really an issue for the kind of applications targetted by this filesystem. I am a bit concerned by the reduction of the CRC from 16 to 8 bits. Given the slot is not there anymore, the way to ensure that the data has been correctly written after a power failure is to check the CRC of the last entry, just like the garbage collector is started again in that case. I don't say it has to be done now, I am not sure if it has to be done in NVS and then fixed or reported to the application, or simply done fully at the application level. What is sure is that we want to have this option. In that regard, I wonder if a CRC8 is not a bit weak. In case the data is not fully written and that the CRC8 byte is still 0xff, it 1 chance over 256 to be correct with the wrong data, which is almost 0.4% and thus not that low. Also note that for a write block size of 2 bytes, and considering all written data is even, using a CRC16 instead of a CRC8 doesn't use more flash space.

I have done some tests on my board, all seems to work fine here. I have just noticed that the garbage collector restart on initialization doesn't work correctly (I reset my board during a garbage collection), it ended up in an endless loop. I don't think it has been introduced by your patch, and can be probably fixed later.

A few other comments reviewing the code:
- You removed the "Initializing non empty sector" check. I think it's worth keeping it, that's how I found the bug in `_nvs_sector_is_used`.
- I guess the `_nvs_flash_cmp` function should also check that the CRC is correct. Otherwise with `CONFIG_NVS_PROTECT_FLASH`, if the CRC ends up being wrong, it's not possible to rewrite the same data. That's something that can be improved later.
- `nvs_get_first_entry`, `nvs_get_last_entry` and `nvs_walk_entry` are relatively similar, I wonder if it would be possible to implement  `nvs_get_first_entry` and `nvs_get_last_entry` by calling `nvs_walk_entry`. Again that's something that can be improved later.

So in short besides the size of the CRC, it looks all good to me. The remaining things can be improved later.

Laczen commented:
Hi @aurel32, thanks for the comments and the testing.

I am still thinking of changing how the data is written to flash, also to
take into account your issue #3 (used space for small items): each item
would be stored as a aligned object: id, length, data followed by the
length again in a multiple of the write_block_size. To further reduce
storage the length would only be of size u8 if it is smaller than 127 bytes
(like in fcb). This would allow up to 4 bytes to be stored in a system with
write_block_size=8. The length is twice written to allow stepping back in
the filesystem and at the same time provides a check if the write was
complete. The CRC would be dropped, but when writing to flash a read back
is performed to verify the write was done properly. If CRC protection of
data against change is required this should be done by the application.

To improve the robustness against powerfail and or flash errors the
filesystem would be locked against writes when there is an error that can
not be corrected (e.g. error when reading back from flash, restarting gc on
init fails, filesystem is full). In this case the application can still
read the existing data, perform a nvs_erase and start working again.

The NVS code should be made leaner, less (redundant) error checking, less
exceptions (e.g. the sector header should become an entry like all other
things in flash), direct reading in flash to step through the filesystem
(either forwards or backwards).

I will make a separate issue on github so that it can be followed.

2018-05-25 13:12 GMT+02:00 aurel32 <notifications@github.com>:

> Hi @Laczen <https://github.com/Laczen>, thanks for your improvements on
> it. To answer quickly your questions about git and PR, in general it's
> better to do small commits that you can reorganize and merge before sending
> a PR. Also it's possible to send patches written by someone else in a PR.
>
> To come back about your patchset, as it changes things that I have
> introduced in my patchset I have just looked at it as a big diff between
> master and your nvs_pr7674. That's probably how things should be submitted
> at this point, a single commit.
>
> The way you bundled the CRC within the padding if possible is very nicely
> done. The only drawback I see is that you can't read the flash if you don't
> know the write block size, but I think it's not really an issue for the
> kind of applications targetted by this filesystem. I am a bit concerned by
> the reduction of the CRC from 16 to 8 bits. Given the slot is not there
> anymore, the way to ensure that the data has been correctly written after a
> power failure is to check the CRC of the last entry, just like the garbage
> collector is started again in that case. I don't say it has to be done now,
> I am not sure if it has to be done in NVS and then fixed or reported to the
> application, or simply done fully at the application level. What is sure is
> that we want to have this option. In that regard, I wonder if a CRC8 is not
> a bit weak. In case the data is not fully written and that the CRC8 byte is
> still 0xff, it 1 chance over 256 to be correct with the wrong data, which
> is almost 0.4% and thus not that low. Also note that for a write block size
> of 2 bytes, and considering all written data is even, using a CRC16 instead
> of a CRC8 doesn't use more flash space.
>
> I have done some tests on my board, all seems to work fine here. I have
> just noticed that the garbage collector restart on initialization doesn't
> work correctly (I reset my board during a garbage collection), it ended up
> in an endless loop. I don't think it has been introduced by your patch, and
> can be probably fixed later.
>
> A few other comments reviewing the code:
>
>    - You removed the "Initializing non empty sector" check. I think it's
>    worth keeping it, that's how I found the bug in _nvs_sector_is_used.
>    - I guess the _nvs_flash_cmp function should also check that the CRC
>    is correct. Otherwise with CONFIG_NVS_PROTECT_FLASH, if the CRC ends
>    up being wrong, it's not possible to rewrite the same data. That's
>    something that can be improved later.
>    - nvs_get_first_entry, nvs_get_last_entry and nvs_walk_entry are
>    relatively similar, I wonder if it would be possible to implement
>    nvs_get_first_entry and nvs_get_last_entry by calling nvs_walk_entry.
>    Again that's something that can be improved later.
>
> So in short besides the size of the CRC, it looks all good to me. The
> remaining things can be improved later.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zephyrproject-rtos/zephyr/pull/7674#issuecomment-392020817>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAjVvPdMHpH5MNdFC_F4Wpb_g4nl52NKks5t1-cAgaJpZM4UF4rU>
> .
>

aurel32 commented:
Hi @Laczen, let me comment that:
> I am still thinking of changing how the data is written to flash, also to take into account your issue #3 (used space for small items): each item would be stored as a aligned object: id, length, data followed by the length again in a multiple of the write_block_size.

One of the thing I have imagined to implement later for NVS is to secure erase some data. This can be done as flash memories can be reprogrammed a second time as long as you write 0x00. The idea was to rewrite the entire data with 0x00. This can not work if the metadata and the data are on the same write block.

> To further reduce storage the length would only be of size u8 if it is smaller than 127 bytes (like in fcb).

Do you still plan to support longer lengths by writing 2 bytes? 127 bytes might be to small for small applications, and I think NVS should be useful for a lot of applications, if we do not want to see it yet another file system.

> The CRC would be dropped, but when writing to flash a read back is performed to verify the write was done properly.

So what is the plan to detect a power failure while the data is being written? Do you plan to use the second length at the end as a verification in `nvs_init`?

>If CRC protection of data against change is required this should be done by the application.

I understand you don't want to force applications to use a CRC, as it clearly takes some space on the flash. Now one of the goal of zephyr is to provide modules that can be used, so that each application does not have to reinvent the wheel. What about a configuration option for that? Either a Kconfig one or an option passed to `nvs_init`?

Anyway do not hesitate to ask me if you want me to review or test your new implementation.

Laczen commented:
Hi @aurel32, I have been working on a rewrite of NVS to provide better flash usage on systems with write block size >4. I have just pushed it to my zephyr fork: https://github.com/Laczen/zephyr/tree/nvs_rework.

There are some big changes in how the data is stored on flash, data is stored in length-id-data-length and this complete set is aligned. The length at the end is used as a verification at start (nvs_init).

A CRC is not added to the data. The main reasoning behind this is that NVS does not know what to do if the CRC check fails. The user application knows best what to do in such a case so it is better to leave the adding of the CRC to the user application.

After each write a read is performed to check if the data was written OK. I am not convinced that this is required.

When an error is detected (either during write/verification/initialization) the filesystem goes into a locked state. This should allow the recovery of data before the error.

As you stated the "secure erase" can not be performed when the data is stored as an aligned structure of length-id-data-length but I think this is a minor disadvantage compared to the advantage of less flash usage.

If you have some time could you test it ? I am especially interested in the result of a powerfail during garbage collection to see if the system starts up in a locked state.
